housing_modeling/src/train.py

```diff
@@ -0,0 +1,201 @@
+        "max_features": randint(low=1, high=8),
+    }
+    forest_reg = RandomForestRegressor(random_state=42)
+    rnd_search = RandomizedSearchCV(
+        forest_reg,
+        param_distributions=param_distribs,
+        n_iter=10,
+        cv=5,
+        scoring="neg_mean_squared_error",
+        random_state=42,
+    )
+    rnd_search.fit(X, y)
+    cvres = rnd_search.cv_results_
+    for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
+        logger.info(
+            "random forest with random search gave score \n %s for parameters %s"
+            % (str(np.sqrt(-mean_score)), str(params))
+        )
+    joblib.dump(
+        rnd_search, os.path.join(MODEL_PATH, "random_forest_random_search.pkl")
+    )
+
+
+def train_RFR_grid_search(X, y):
+    """
+    Trains a random forest model on the inputs using grid search and saves
+    the model as a pickle file
+
+    Parameters:
+    X (pd.DataFrame): training data
+    y (pd.Series):  value to be predicted
+    """
+    param_grid = [
+        # try 12 (3×4) combinations of hyperparameters
+        {"n_estimators": [3, 10, 30], "max_features": [2, 4, 6, 8]},
+        # then try 6 (2×3) combinations with bootstrap set as False
+        {
+            "bootstrap": [False],
+            "n_estimators": [3, 10],
+            "max_features": [2, 3, 4],
+        },
+    ]
+
+    forest_reg = RandomForestRegressor(random_state=42)
+    # train across 5 folds, that's a total of (12+6)*5=90 rounds of training
+    grid_search = GridSearchCV(
+        forest_reg,
+        param_grid,
+        cv=5,
+        scoring="neg_mean_squared_error",
+        return_train_score=True,
+    )
+    grid_search.fit(X, y)
+
+    grid_search.best_params_
+    cvres = grid_search.cv_results_
+    for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
+        logger.info(
+            "random forest with grid search gave score \n %s for parameters %s"
+            % (str(np.sqrt(-mean_score)), str(params))
+        )
+
+    feature_importances = grid_search.best_estimator_.feature_importances_
+    sorted(zip(feature_importances, X.columns), reverse=True)
+
+    final_model = grid_search.best_estimator_
+    joblib.dump(
+        final_model, os.path.join(MODEL_PATH, "random_forest_grid_search.pkl")
+    )
+
+
+if __name__ == "__main__":
+    strat_train_set = pd.read_parquet(
+        os.path.join(HOUSING_PATH, "train.parquet")
+    )
+
+    housing = strat_train_set.copy()
+    logger = create_logger(LOGGING_PATH, "train.log")
+    corr_matrix = housing.corr()
+    corr_matrix["median_house_value"].sort_values(ascending=False)
+    housing["rooms_per_household"] = (
+        housing["total_rooms"] / housing["households"]
+    )
+    housing["bedrooms_per_room"] = (
+        housing["total_bedrooms"] / housing["total_rooms"]
+    )
+    housing["population_per_household"] = (
+        housing["population"] / housing["households"]
+    )
```

Comment:
these new feature computations are repeated in multiple places, (train and test). It would be good to move them to a function and use the function instead.